lib/src/main/java/graphql/nadel/engine/blueprint/NadelVirtualTypeBlueprintFactory.kt

```diff
@@ -0,0 +1,182 @@
+                    }
+            }
+            .toList()
+    }
+
+    fun makeVirtualTypeContext(
+        engineSchema: GraphQLSchema,
+        containerType: GraphQLObjectType,
+        virtualFieldDef: GraphQLFieldDefinition,
+    ): NadelVirtualTypeContext? {
+        val factoryContext = FactoryContext(
+            engineSchema,
+        )
+
+        return with(factoryContext) {
+            makeVirtualTypeContext(containerType, virtualFieldDef)
+        }
+    }
+
+    context(FactoryContext)
+    private fun makeVirtualTypeContext(
+        containerType: GraphQLObjectType,
+        virtualFieldDef: GraphQLFieldDefinition,
+    ): NadelVirtualTypeContext? {
+        val typeMappings = createTypeMappings(virtualFieldDef)
+
+        return if (typeMappings.isEmpty()) {
+            return null
+        } else {
+            NadelVirtualTypeContext(
+                virtualFieldContainer = containerType,
+                virtualField = virtualFieldDef,
+                virtualTypeToBackingType = typeMappings.virtualTypeToBackingTypeMap(),
+                backingTypeToVirtualType = typeMappings.backingTypeToVirtualTypeMap(),
+            )
+        }
+    }
+
+    context(FactoryContext)
+    private fun createTypeMappings(
+        virtualFieldDef: GraphQLFieldDefinition,
+    ): List<VirtualTypeMapping> {
+        val hydration = virtualFieldDef.getHydratedOrNull()
+            ?: return emptyList() // We should use the non-null method once we delete @hydratedFrom
+        val backingFieldDef = engineSchema.queryType.getFieldAt(hydration.backingField)!!
+        val backingType = backingFieldDef.type.unwrapAll() as? GraphQLObjectType
+            ?: return emptyList()
+        val virtualType = virtualFieldDef.type.unwrapAll() as? GraphQLObjectType
+            ?: return emptyList()
+
+        // Not a virtual type
+        if (virtualType.name == backingType.name) {
+            return emptyList()
+        }
+
+        return createTypeMappings(
+            backingType = backingType,
+            virtualType = virtualType,
+        )
+    }
+
+    context(FactoryContext)
+    private fun createTypeMappings(
+        backingType: GraphQLObjectType,
+        virtualType: GraphQLObjectType,
+    ): List<VirtualTypeMapping> {
+        return listOf(VirtualTypeMapping(virtualType, backingType)) + createTypeMappings(
+            virtualType = virtualType,
+            backingType = backingType,
+            visitedVirtualTypes = mutableSetOf(virtualType.name),
+        )
+    }
+
+    context(FactoryContext)
+    private fun createTypeMappings(
+        virtualType: GraphQLObjectType,
+        backingType: GraphQLObjectType,
+        visitedVirtualTypes: MutableSet<String>,
+    ): List<VirtualTypeMapping> {
+        return virtualType
+            .fields
+            .flatMap { virtualFieldDef ->
+                val virtualOutputType = virtualFieldDef.type.unwrapAll() as GraphQLObjectType
```

Comment:
I'm pretty sure this will blow up if the virtual type has any scalar field, like the pagination fields etc. The tests should cover that.

Reply:
Good point

Reply:
Added a basic test to cover, will add more once I merge the validation + blueprint code.